_posts/2015-03-23-design-patterns-part-3.markdown

```diff
@@ -0,0 +1,241 @@
+
+
+The factory is the lynchpin for the whole operation.  That’s where we all of our contexts get created with their real, live, collaborator classes (as opposed to the mocks we used in testing).
+
+Now that our controller has an instance of our CreateUser context in it's handsl, let's take a look at what code is executed when the controller calls our context. In this case our business logic consists of creating a user in the database and sending
+a “welcome” email. Let's take a look at what code is executed when the controller calls our context. 
+
+
+{% highlight ruby %}
+
+class CreateUser
+  def initialize(user_repository, mailer)
+    user_repository = repo
+    mailer = mailer
+  end
+
+  def call(first_name, last_name, email)
+    user = user_repository.create(first_name, last_name, email)
+    mailer.welcome_email(user.first_name, user.email)
+    user
+  end
+
+  private
+ 
+  attr_reader :user_repository, :mailer
+
+end
+
+{% endhighlight %}
+
+Looking inside the call method, we can see that our context uses our repository, which has a pretty generic method called “create.”  The context expects that calling the "create" method will return some
+object that it can call "first_name" and "email" on.  We know that object is our custom entity class. 
+
+
+{% highlight ruby %}
+
+class ARUserRepository
+
+  def initialize(user_model, user_entity)
+    @user_model = user_model
+    @user_entity = user_entity
+  end
+
+  def create(first_name, last_name, email)
+    ar_user_object = User.create!(first_name: first_name, last_name: last_name, email: email)
+    user = translate_record(ar_user_object)
+    user
+  end
+
+  private
+ 
+  attr_reader :user_model, :user_entity
+
+  def translate_record(record)
+    user_entity.new(record.attributes)
+  end
+
+end
+
+{% endhighlight %}
+
+Our call chain starts in the controller, which relies on the factory to create an instance of our context.  Our controller then calls that context, being sure to pass it the appropriate arguments.
+Inside our context, we pass along the user attributes to our repository.  Our repository relies on an ORM (in this case, ActiveRecord) to create the record in the database and we take the result of that process and wrap it into our custom entity.  
+
+##Why?
+
+So even if we understand this elaborate call chain, we may have one lingering question, which is “Why would we ever to through all that trouble when Rails makes it so much easier to create a user?” 
+
+The answer is: it depends. We might want it.  We might not want it.
+
+Whether it makes sense to use all these patterns together depends on what you’re building.  If you’re in a 24 hour hackathon, should just write your
+code as quickly as possible.  You shouldn't worry about all these design pattens.  Or, if you’re taking a few days to crank out an MVP, then go
+ahead of an just write an idiomatic Rails app.  Once you’ve validated your idea, then it's time to think about how you're going to architect your
```

Comment:
`Or, if you’re taking a few days to crank out an MVP, then go ahead of an just write an idiomatic Rails app.` is awkward. Can you fix this up?